Sources/Plasma/FeatureLib/pfGLPipeline/plGLDevice.cpp

```diff
@@ -40,29 +40,891 @@ You can contact Cyan Worlds, Inc. by email legal@cyan.com
+            break;
+        }
+
+        if (CGLSetCurrentContext(ctx) != kCGLNoError) {
+            dev->fErrorMsg = "Failed to attach CGL context to surface";
+            break;
+        }
+
+        CGLReleasePixelFormat(pix);
+
+        // Successfully initialized:
+        dev->fContext = ctx;
+        dev->fContextType = plGLDevice::kCGL;
+        return;
+    } while (0);
+
+    // Cleanup for failure case:
+    if (ctx) {
+        CGLSetCurrentContext(nullptr);
+        CGLReleaseContext(ctx);
+    }
+
+    if (pix)
+        CGLReleasePixelFormat(pix);
+
+    IGNORE_WARNINGS_END
+}
+
+void FiniCGLDevice(plGLDevice* dev)
+{
+    IGNORE_WARNINGS_BEGIN("deprecated-declarations")
+
+    CGLContextObj ctx = static_cast<CGLContextObj>(dev->fContext);
+
+    if (ctx) {
+        CGLSetCurrentContext(nullptr);
+        CGLReleaseContext(ctx);
+    }
+
+    IGNORE_WARNINGS_END
+}
+#endif // HS_BUILD_FOR_MACOS
+#pragma endregion CGL_Init
+
+#ifdef HS_DEBUGGING
+void GLAPIENTRY plGLDebugLog(GLenum source, GLenum type, GLuint id, GLenum severity, GLsizei length, const GLchar* message, const void* userParam)
+{
+    if (severity <= GL_DEBUG_SEVERITY_MEDIUM) { // Yes, higher is a lower enum value
+        plStatusLog::AddLineSF("pipeline.log", "[GL] {}{}", (type == GL_DEBUG_TYPE_ERROR ? "** ERROR **: " : ""), message);
+    }
+}
+#endif
+
+static float kIdentityMatrix[16] = {
+    1.0f, 0.0f, 0.0f, 0.0f,
+    0.0f, 1.0f, 0.0f, 0.0f,
+    0.0f, 0.0f, 1.0f, 0.0f,
+    0.0f, 0.0f, 0.0f, 1.0f
+};
+
+GLfloat* hsMatrix2GL(const hsMatrix44& src, GLfloat* dst)
+{
+    if (src.fFlags & hsMatrix44::kIsIdent)
+        return static_cast<GLfloat*>(memcpy(dst, kIdentityMatrix, sizeof(GLfloat) * 16));
+    else
+        return static_cast<GLfloat*>(memcpy(dst, src.fMap, sizeof(GLfloat) * 16));
+}
 
 plGLDevice::plGLDevice()
-    : fErrorMsg()
+    : fErrorMsg(), fPipeline(), fContextType(kNone), fWindow(), fDevice(),
+    fDisplay(), fSurface(), fContext(), fActiveThread(), fCurrentProgram()
+{
+    memcpy(fMatrixL2W, kIdentityMatrix, sizeof(GLfloat) * 16);
+    memcpy(fMatrixW2C, kIdentityMatrix, sizeof(GLfloat) * 16);
+    memcpy(fMatrixProj, kIdentityMatrix, sizeof(GLfloat) * 16);
+}
+
+bool plGLDevice::InitDevice()
+{
+#ifdef USE_EGL
+    // The USE_EGL define tells us whether the epoxy library includes support
+    // for attempting to use EGL on the current platform, but we still need to
+    // check if EGL is actually available at runtime.
+    //
+    // On Windows, this may be true in cases like the PowerVR SDK or when using
+    // ANGLE.
+    //
+    // On Linux, this should be true with mesa or nvidia drivers.
+    if (epoxy_has_egl() && fContextType == kNone)
+        InitEGLDevice(this);
+#endif
+
+#ifdef HS_BUILD_FOR_WIN32
+    if (fContextType == kNone)
+        InitWGLDevice(this);
+#endif
+
+#ifdef HS_BUILD_FOR_MACOS
+    if (fContextType == kNone)
+        InitCGLDevice(this);
+#endif
```

Comment:
This really feels like a good opportunity for some virtual functions for each context type.